app/src/main/java/com/speakeasy/watsonbarassistant/MainMenu.kt

```diff
@@ -129,3 +177,102 @@ class MainMenu : AppCompatActivity() {
+        override fun compare(a: DiscoveryRecipe, b: DiscoveryRecipe): Int = when {
+            a.queueValue < b.queueValue -> 1
+            else -> -1
+        }
+    }
+}
+
+ class SearchDiscovery(inputListener: OnTaskCompleted) : AsyncTask<List<Ingredient>, Void, MutableList<DiscoveryRecipe>>() {
+
+     private var listener: OnTaskCompleted
+
+     init {
+         this.listener = inputListener
+     }
+
+    override fun doInBackground(vararg args: List<Ingredient>): MutableList<DiscoveryRecipe> {
+        var ingredients = args[0]
+        var orderedRecipes = PriorityQueue<DiscoveryRecipe>(10, CompareLists)
+
+        Log.i("Discovery", "Start.")
+        val discovery = Discovery(
+                VERSION_DIS,
+                USERNAME_MIKE_DIS,
+                PASSWORD_MIKE_DIS
+        )
+        discovery.endPoint = URL_MIKE_DIS
+
+        val queryBuilder = QueryOptions.Builder(ENV_ID_MIKE_DIS, COL_ID_MIKE_DIS)
+
+        queryBuilder.query(buildIngredientQuery(ingredients)).count(50)
+        val queryResponse = discovery.query(queryBuilder.build()).execute()
+
+        Log.i("Discovery", "Serialization Start.")
+        //Log.i("Discovery", "Response: " + queryResponse.results[0])
+        for(response in queryResponse.results){
+            val recipe = JSON.nonstrict.parse<DiscoveryRecipe>(response.toString())
+            if(recipe.imgageBase64 == ""){
+                Log.i("Discovery","Failed: " + recipe.title + ": " + response.toString() + "\n")
+            }
+            recipe.calculatePercentAvailable(ingredients)
+            orderedRecipes.add(recipe)
+        }
+
+        Log.i("Discovery", "Serialization Finished.")
+
+        return orderedRecipes.toMutableList()
+    }
+
+    protected fun onProgressUpdate(vararg progress: Int) {
+    }
+
+    override fun onPostExecute(result: MutableList<DiscoveryRecipe>){
+        super.onPostExecute(result)
+
+        this.listener.onTaskCompleted(result)
+    }
+
+    private fun buildIngredientQuery(ingredients: List<Ingredient>): String{
+        val queryString = "ingredientList:"
+        val ingredients = ingredients
+        val sb = StringBuilder()
+        sb.append(queryString)
+
+        for (ingredient in ingredients){
+            if(ingredient.name != "") {
+                sb.append(ingredient.name)
+                sb.append("|")
+            }
+        }
+
+        sb.replace(sb.length-1,sb.length, "")
+        Log.i("Discovery", "Ingredient Query: " + sb.toString())
+        return sb.toString()
+    }
+}
+
+interface OnTaskCompleted {
+    fun onTaskCompleted(recipes: MutableList<DiscoveryRecipe>)
+}
+
+class HandleDiscovery(overAllList: MutableList<MutableList<DiscoveryRecipe>>): OnTaskCompleted{
+    private var list = mutableListOf<MutableList<DiscoveryRecipe>>()
+    init{
+        this.list = overAllList
+    }
+    override fun onTaskCompleted(recipes: MutableList<DiscoveryRecipe>) {
+        this.list.add(0, recipes)
+        this.list.add(1, recipes)
```

Comment:
Maybe call 
```kotlin
recipes.shuffle()
```
on the second add so the recipes won't be displayed in the same order twice.